src/de.rs

```diff
@@ -398,6 +400,50 @@ impl<S: Read> Decoder<S> {
         Ok(i)
     }
 
+    pub(crate) async fn peek(&mut self) -> Result<Option<u8>, Error> {
+        while self.buffer.is_empty() && !self.source.is_terminated() {
+            self.buffer().await?;
+        }
+
+        if self.buffer.is_empty() {
+            Ok(None)
+        } else {
+            Ok(Some(self.buffer[0]))
+        }
+    }
+
+    async fn peek_or_null(&mut self) -> Result<u8, Error> {
+        Ok(self.peek().await?.unwrap_or(b'\x00'))
+    }
+
+    async fn next_char(&mut self) -> Result<Option<u8>, Error> {
+        while self.buffer.is_empty() && !self.source.is_terminated() {
+            self.buffer().await?;
+        }
+
+        Ok(Some(self.buffer.remove(0)))
+    }
+
+    async fn eat_char(&mut self) -> Result<(), Error> {
+        self.next_char().await?;
+        Ok(())
+    }
+
+    async fn next_char_or_null(&mut self) -> Result<u8, Error> {
+        Ok(self.next_char().await?.unwrap_or(b'\x00'))
```

Comment:
Same here, wouldn't it be better to check if `next_char` is `None` than to return a null?